### PR TITLE
Fix problem with wrong closure and add possibility to blacklist other plugins.

### PR DESCRIPTION
**Pull request by Szymon20000**
#### WHY:
Some of the plugins that are used by metro-react-native-preset or by babel itself don't update correctly bindings properties in scopes. Because we use bindings in order to find all the missing variables to capture them we have to work with scopes that reflect a current state in AST.

Example:
``` JS
let {x, y} = obj;
``` 
is changed to 
```JS
let _obj = obj,
    x = _obj.x,
    y = _obj.y;
```
by babel-plugin-transform-destructuring.
However, in binding, we only find [x, y].

#### SOLUTION
Work on a copy of worklet tree with proper bindings:
``` JS
const copyWithCorrentBindings = Parser.parse('(' + fun.toString() + '\n)');
traverse(copyWithCorrentBindings, {...visitors}); 
```
----------------------------------------------------
#### WHY:
Most of the babel presets used by React Native developers contains plugins that polyfills mechanisms that are supported by Hermes and JSC. Additionally, some of the polyfills use external code that cannot be captured by worklets. 

Example:
Object.assign polyfill
```JS
Object.assign(a, b);
```
is changed to 
``` JS
var _extends = ...;
_extends(a, b);
```
by  @babel/plugin-transform-object-assign.

#### SOLUTION
There is something like ```manipulateOptions``` property that plugin can use to access babel configuration. Unfortunately, when a plugin is added within a preset plugin name is replaced by file descriptor name that says nothing. In order to find which plugin should be removed, we get plugin visitors from node_modules and then compare keys. In most cases (always) that step is enough. However, to be sure we additionally compare visitor string representation (but only if keys are the same).

## TESTING
Check [example](https://github.com/software-mansion/react-native-reanimated/pull/915/files)
Check also if Object.assign is visible in stylediff worklet.

## CLOSING
resolves: https://github.com/software-mansion/react-native-reanimated/issues/921
resolves: https://github.com/software-mansion/react-native-reanimated/issues/939

### CODE REVIEW

**Comment by wcandillon** [comment]
Thank you @Szymon20000 🙌🏻 Should we merge https://github.com/software-mansion/react-native-reanimated/pull/915 as a test or close it maybe?

**Comment by Szymon20000** [comment]
It's a good test and I think we should have it but I'm not sure if we want it as an example that people should see. It should be more like a test screen. I think we need another directory with test screens. @karol-bisztyga  What do you think?

File: plugin.js [hunk 1]
```diff
@@ -281,24 +286,64 @@ function processWorklets(t, path, processor) {
...
+});
+
+function removePluginsFromBlacklist(plugins) {
+  PLUGIN_BLACKLIST.forEach((blacklistedPlugin) => {
+    if (!blacklistedPlugin) {
+      return;
+    }
+
+    let toRemove = [];
+    for (let i = 0; i < plugins.length; i++) {
```
Comment (karol-bisztyga) [inline plugin.js:315]: I think it could be done using something like `Array.filter`

File: plugin.js [hunk 2]
```diff
@@ -281,24 +286,64 @@ function processWorklets(t, path, processor) {
...
+      if (
+        JSON.stringify(Object.keys(plugins[i].visitor)) !=
+        JSON.stringify(Object.keys(blacklistedPlugin.visitor))
+      ) {
+        continue;
+      }
+      let areEqual = true;
```
Comment (karol-bisztyga) [inline plugin.js:322]: Can be determined using `Array.some`

**Comment by karol-bisztyga** [comment]
> It's a good test and I think we should have it but I'm not sure if we want it as an example that people should see. It should be more like a test screen. I think we need another directory with test screens. @karol-bisztyga What do you think?

Yes this is definitely a good idea, I was thinking about it too!

**Comment by wcandillon** [comment]
@Szymon20000 alternatively, I can update the example as a nice example people could see. It just wanted to show the minimal reproducible example. Let me know. Thank you for much for looking into it ♥️

File: plugin.js [hunk 3]
```diff
@@ -90,14 +93,11 @@ function processWorkletFunction(t, fun) {
...
+  // some even brake them!
+  const astWorkletCopy = parse('\n(' + fun.toString() + '\n)');
 
-  fun.traverse({
-    DirectiveLiteral(path) {
-      if (path.node.value === 'worklet' && path.getFunctionParent() === fun) {
-        path.parentPath.remove();
-      }
-    },
+  traverse(astWorkletCopy, {
```
Comment (kmagiera) [inline plugin.js:100]: Wouldn't it be safer to extract "outputs" on the parsed copy instead of keeping it in `fun.traverse` block? Then we can keep `fun.traverse` only to do `'worklet'` directive cleanup

File: plugin.js [hunk 4]
```diff
@@ -281,24 +286,64 @@ function processWorklets(t, path, processor) {
   }
 }
 
+const PLUGIN_BLACKLIST_NAMES = [
+  // '@babel/plugin-transform-destructuring', do not blacklist
```
Comment (kmagiera) [inline plugin.js:290]: why is this here. if its not on a list, there is no point of adding it in a comment as if it "almost" was on a list

File: plugin.js [hunk 5]
```diff
@@ -281,24 +286,64 @@ function processWorklets(t, path, processor) {
...
+  'metro-react-native-babel-preset/node_modules/@babel/plugin-transform-object-assign',
+];
+
+const PLUGIN_BLACKLIST = PLUGIN_BLACKLIST_NAMES.map((pluginName) => {
+  try {
+    const blacklistedPluginObject = require(pluginName);
+    const blacklistedPlugin = blacklistedPluginObject.default({
```
Comment (kmagiera) [inline plugin.js:297]: this seem to be very obscure and rely on undocumented babel features. Can we add some comments on what it does as when the babel api changes we won't be able to track down what this piece of code should do

File: plugin.js [hunk 6]
```diff
@@ -281,24 +286,64 @@ function processWorklets(t, path, processor) {
...
+      assertVersion: (x) => true,
+    });
+
+    visitors.explode(blacklistedPlugin.visitor);
+    return blacklistedPlugin;
+  } catch (e) {
+    console.log(`No ${pluginName}!`);
```
Comment (kmagiera) [inline plugin.js:304]: please remove console.log statements. If this is an error let's throw here instead? Or use console.warn or anything that'd suggest you are not accidently commiting some debug/logging code

File: plugin.js [hunk 7]
```diff
@@ -281,24 +286,64 @@ function processWorklets(t, path, processor) {
...
+function removePluginsFromBlacklist(plugins) {
+  PLUGIN_BLACKLIST.forEach((blacklistedPlugin) => {
+    if (!blacklistedPlugin) {
+      return;
+    }
+
+    let toRemove = [];
+    for (let i = 0; i < plugins.length; i++) {
+      if (
+        JSON.stringify(Object.keys(plugins[i].visitor)) !=
```
Comment (kmagiera) [inline plugin.js:317]: shouldn't those objects just be the same? we load plugins using require and I think there should always be one source where plugins are loaded, no?

**Comment by Szymon20000** [comment]
> Nice to see a more resilient solution to typescript and related issues we were having. Left a few comments inline

I think that this is a proper solution. The previous one was wrong. Our plugin has to process worklets after all other plugins.  People can use their own plugins and we want to let them do that.

File: plugin.js (hunk 7 above)
Comment (Szymon20000) [reply plugin.js:317]: Not true. Plugins can be loaded by several presets. Moreover, "require" returns a function, not an object.

File: plugin.js [hunk 8]
```diff
@@ -90,14 +93,11 @@ function processWorkletFunction(t, fun) {
   const closure = new Map();
   const outputs = new Set();
 
-  const funScope = fun.scope;
+  // We use copy because some of the plugins don't update bindings and
+  // some even brake them!
```
Comment (kmagiera) [inline plugin.js:97]:
```suggestion
  // some even break them
```

File: plugin.js [hunk 9]
```diff
@@ -308,5 +354,9 @@ module.exports = function ({ types: t }) {
         },
       },
     },
+    manipulateOptions(opts, parserOpts) {
```
Comment (kmagiera) [inline plugin.js:357]: Can we document here what this thing does? Babel documentsation is very hard to browse and I don't understand and which point the options are going to be changed. Does the chaneg apply only to the AST branches that we traverse? Or maybe it is applied on a single-file level? Or maybe it is applied to all the files? Why we don't restore the previous options?

File: plugin.js (hunk 7 above)
Comment (kmagiera) [reply plugin.js:318]: I wish we'd used soem different approach than JSON serialization. There is nothing technically wrong with it but it feels like the plugin object should have some field that we can extract its name from? If we want to use this as a last resort there should be a proper comment explaining that plugin objects are created by babel internals and they don't carry any identifier etc...

File: plugin.js (hunk 9 above)
Comment (Szymon20000) [reply plugin.js:359]: It impacts all files. We don't have to restore previous options because Hermes and JSC support Object.assign.

**Comment by wcandillon** [comment]
Congratulations 🙌🏻
